File: source/backend/src/main/java/com/defecttriage/entity/Defect.java
package com.defecttriage.entity;

import com.defecttriage.common.DefectStatus;
import com.defecttriage.common.Priority;
import jakarta.persistence.*;
import lombok.Getter;
import lombok.NoArgsConstructor;
import lombok.Setter;

import java.time.LocalDateTime;

@Entity
@Table(name = "defects")
@Getter @Setter @NoArgsConstructor
public class Defect {
    @Id @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    @Column(nullable = false)
    private String title;

    @Column(columnDefinition = "TEXT")
    private String description;

    @Column(columnDefinition = "TEXT")
    private String phenomenon;

    @Column(columnDefinition = "TEXT")
    private String environment;

    @Column(columnDefinition = "TEXT")
    private String reproductionSteps;

    @Column(columnDefinition = "TEXT")
    private String expectedResult;

    @Column(columnDefinition = "TEXT")
    private String actualResult;

    private Integer severity;

    @Enumerated(EnumType.STRING)
    private Priority priority;

    private Integer userImpact;
    private Integer businessImpact;
    private Integer frequency;
    private Integer workaround;
    private Integer releaseWindow;

    @Column(columnDefinition = "TEXT")
    private String rootCauseHypothesis;

    @Column(columnDefinition = "TEXT")
    private String fixPlan;

    @Column(columnDefinition = "TEXT")
    private String fixContent;

    private String affectedModules;

    private Integer fixDuration;

    @Column(columnDefinition = "TEXT")
    private String verificationResult;

    @Column(columnDefinition = "TEXT")
    private String regressionScope;

    @Column(columnDefinition = "TEXT")
    private String verificationConclusion;

    @Enumerated(EnumType.STRING)
    @Column(nullable = false)
    private DefectStatus status = DefectStatus.DRAFT;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "reporter_id")
    private User reporter;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "assignee_id")
    private User assignee;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "verifier_id")
    private User verifier;

    private LocalDateTime createdAt;
    private LocalDateTime updatedAt;
    private LocalDateTime closedAt;

    @Version
    private Long version;

    @PrePersist
    void prePersist() {
        createdAt = LocalDateTime.now();
        updatedAt = LocalDateTime.now();
    }

    @PreUpdate
    void preUpdate() {
        updatedAt = LocalDateTime.now();
    }
}

File: source/frontend/src/types/defect.ts
export type DefectStatus =
  | 'DRAFT' | 'REPORTED' | 'TRIAGING' | 'ANALYZED'
  | 'PLANNED' | 'IN_REPAIR' | 'FIXED' | 'VERIFIED' | 'CLOSED' | 'REOPENED'

export type Priority = 'P0' | 'P1' | 'P2' | 'P3' | 'P4'

export interface DefectListItem {
  id: number
  title: string
  status: DefectStatus
  priority: Priority | null
  severity: number | null
  reporterName: string | null
  assigneeName: string | null
  createdAt: string
}

export interface Defect {
  id: number
  title: string
  description: string | null
  phenomenon: string | null
  environment: string | null
  reproductionSteps: string | null
  expectedResult: string | null
  actualResult: string | null
  severity: number | null
  priority: Priority | null
  userImpact: number | null
  businessImpact: number | null
  frequency: number | null
  workaround: number | null
  releaseWindow: number | null
  rootCauseHypothesis: string | null
  fixPlan: string | null
  fixContent: string | null
  affectedModules: string | null
  fixDuration: number | null
  verificationResult: string | null
  regressionScope: string | null
  verificationConclusion: string | null
  status: DefectStatus
  reporterId: number | null
  reporterName: string | null
  assigneeId: number | null
  assigneeName: string | null
  verifierId: number | null
  verifierName: string | null
  createdAt: string
  updatedAt: string
  closedAt: string | null
}

export interface CreateDefectRequest {
  title: string
  description?: string
  phenomenon?: string
  environment?: string
  reproductionSteps?: string
  expectedResult?: string
  actualResult?: string
}

export interface DefectUpdateRequest {
  title?: string
  description?: string
  phenomenon?: string
  environment?: string
  reproductionSteps?: string
  expectedResult?: string
  actualResult?: string
  severity?: number
  priority?: Priority
  userImpact?: number
  businessImpact?: number
  frequency?: number
  workaround?: number
  releaseWindow?: number
  rootCauseHypothesis?: string
  fixPlan?: string
  fixContent?: string
  affectedModules?: string
  fixDuration?: number
  verificationResult?: string
  regressionScope?: string
  verificationConclusion?: string
}

export interface StateTransition {
  id: number
  fromStatus: DefectStatus
  toStatus: DefectStatus
  operatorName: string | null
  note: string | null
  createdAt: string
}
